sdk/offchain/src/identifier.rs

```diff
@@ -0,0 +1,386 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::subaddress::{Subaddress, SubaddressParseError};
+use anyhow::{anyhow, Result};
+use bech32::{self, u5, FromBase32, ToBase32};
+use diem_sdk::{
+    move_types::account_address::AccountAddressParseError,
+    types::{
+        account_address::AccountAddress,
+        account_config::allowed_currency_code_string,
+        chain_id::{ChainId, NamedChain},
+    },
+};
+use std::collections::HashMap;
+use thiserror::Error;
+use url::Url;
+
+const DIEM_BECH32_VERSION: u8 = 1;
+
+/// Intent is a struct holdind data decoded from Diem Intent Identifier string
+/// https://dip.diem.com/dip-5/#format
+#[derive(Ord, PartialOrd, Eq, PartialEq, Hash, Clone)]
+pub struct Intent {
+    pub account_address: AccountAddress,
+    pub subaddress: Subaddress,
+    pub currency_code: Option<String>,
+    pub amount: Option<u64>,
+    pub hrp: String,
+}
+
+impl Intent {
+    /// Decode Diem intent identifier (https://dip.diem.com/dip-5/) int 3 parts:
+    /// 1. account identifier: account address & sub-address
+    /// 2. currency code
+    /// 3. amount
+    ///
+    /// ## Example
+    /// ```
+    /// use offchain::identifier::Intent;
+    /// let identifier = "diem://dm1p7ujcndcl7nudzwt8fglhx6wxn08kgs5tm6mz4us2vfufk?c=XUS&am=4500";
+    /// let intent = Intent::from(identifier, "dm").unwrap();
+    /// assert_eq!(intent.hrp, "dm");
+    /// assert_eq!(intent.amount, Some(4500));
+    /// ```
+    pub fn from(
+        encoded_intent_identifier: &str,
+        hrp: &str,
```

Comment:
I don't know if there's much value with including the expected hrp here. This type should define everything you need to work with Intents but then user's of this can do the check's themselves to ensure that the hrp is what they expect.